Move most world/debug options to externals

RenechCDDA opened the pull request:
#### Summary
None

#### Purpose of change
We have a plethora of world options. Too many. This is an opinion that has been repeatedly stated by senior developers, and one that I share.

Presenting these options to players implies that it is intended for players to casually fiddle with them. It is not. These are major balance levers, meant for dedicated mods.

#### Describe the solution
This PR is opened as a draft, so the solution describes how it should be when it's done. It is not done at the time of this writing.

DEBUG:
Distance initial visibility ---> Moved to scenario-specific value. Same default as before, but now scenarios can twiddle with it.

Vertical Range of 3D field of vision ---> Option removed. No reason to keep it around, it should have gone with the option to disable 3D FOV.  Value moved to game constants and locked at 4.

WORLD:
Size of cities ---> Untouched (for now), retained on the main menu
City spacing ---> Untouched (for now), retained on the main menu

Difficulty (slider) ---> Removed, underlying options moved to externals
Spawn rate/item scaling factor --> The underlying options in question
Monster speed/monster resilience --> The underlying options in question

Random NPCs (slider) ---> Removed, underlying option moved to externals
Random NPC spawn time --> The underlying option in question

Monster evolution slowdown --> Moved to external

Season length --> Moved to external
Construction scaling --> Moved to external
Eternal Season --> Moved to external

Wandering Hordes --> Untouched for now, option will be removed when hordes are in a better state. (That's been the plan for a long time, tangential to this PR)

Mutations by radiation ---> Moved to external

#### Describe alternatives you've considered


#### Testing
Opened as draft

#### Additional context

Maleclypse commented:
I would reverse all of the following. 

```
WORLD:
Size of cities ---> Untouched (for now), retained on the main menu
City spacing ---> Untouched (for now), retained on the main menu

Difficulty (slider) ---> Removed, underlying options moved to externals
Monster speed/monster resilience --> The underlying options in question

Random NPCs (slider) ---> Removed, underlying option moved to externals
Random NPC spawn time --> The underlying option in question

Monster evolution slowdown --> Moved to external
```

Size of cities and city spacing should both be moved to external only as both of these break special generation in ways that aren't easily addressable at all and should exclusively be something that exists for innawoods or similar mods.  It should never be used in any mod that isn't a total conversion mod and I'd consider it ideal with it forcing that kind of labeling in the mod menu even for third party mods.
Difficulty slider could be used to move monster speed and resilience up or down all at once or both could remain as independent sliders but some kind of difficulty adjustment should exist for players who only play whatever ships with the game.  If we get rid of this slider we will have to ship difficulty adjustment mods with the game that basically exist only for that one purpose and it seems silly.  
I did remove item spawn slider from this list as I agree it makes management of mod spawns difficult to control for.  
NPC slider is nice to have because if you move it exclusively to externals I'm going to straight jack it up in XE because I think that world would have more survivors due to longer weirdness and how will that interact if some other mod drops it down but only XE mentions what it's doing.  I'd rather just leave this in player control as it only affects random npcs.

Monster Evolution I think should stay but I don't have any real arguments for it other than I think it would worsen the experience for players who are unaware of github and third party modding especially if we didn't then ship a couple mod options around this with the game which again I believe would just clog up the mod menu.  Also how would mods interact with each others' slider adjustments to this?

esotericist commented:
> If we get rid of this slider we will have to ship difficulty adjustment mods with the game that basically exist only for that one purpose and it seems silly.

we have no such obligation.

Maleclypse commented:
> > If we get rid of this slider we will have to ship difficulty adjustment mods with the game that basically exist only for that one purpose and it seems silly.
> 
> we have no such obligation.

You have no obligations.

John-Candlebury commented:
I agree with Maleclypse assesment of the options, I think theres place for some small but impactful levers to difficulty in user facing settings, and the ones outlined in his post are effective and dont really seem to cause problems with development.

harakka commented:
> Presenting these options to players implies that it is intended for players to casually fiddle with them. It is not.

Back when the current simplified world settings screen was added a couple of stables ago, the current difficulty slider was specifically created as a new thing so that there is something simple for the player to adjust difficulty with that doesn't break the game like a lot of the other settings can do.

IdleSol commented:
*I forbid you to not play the way I want you to play. ... I miscalculated, but where?*

Maddremor commented:
Earnest question: Does tweaking monster evolution have any knock-on effects that are a hassle to deal with? As difficulty goes, I think it's broadly reasonable to have the game ramp up slower for a less skilled player. Monster evolution isn't pegged to any milestones or time frames that are grounded in reality, right? If it messes with wildlife spawns or such then do tell me.

kevingranade commented:
@IdleSol you have made several heckling comments like this recently in different issues. This is not ok and if repeated will result in a permanent ban.

RenechCDDA commented:
So my main reason for not removing the city size/spacing slider was that there would be no sliders left. It would be... weird.

Reviewing the points brought up, it seems most of the design conversation around the existing sliders was #57098. I'm really not *convinced* that any sliders should exist, but I don't have very strong arguments against them at the moment. What I do take issue with is the city size slider. Ironically, that is the one I most think should be removed. As mentioned, it is by far the worst offender of breaking the game.

So my current plan, in light of feedback, is to reverse course on the sliders.

City size/spacing slider ---> Removed, options moved to externals

Difficulty/NPC sliders ---> Kept

>Also how would mods interact with each others' slider adjustments to this?

Any external options would be loaded like other json objects, with the last loaded mod taking priority. e.g. if evolution was an external and someone loaded two mods with both defining different values then the last one in the mod list is applied.

All the sliders are currently 'main options', so only one thing can set them. (The player, via the options menu or the sliders during world creation).

I am not 100% sure on how I could/would move the slider options to externals while still keeping them modifiable by sliders, but that's part of the working problem here.

RenechCDDA committed:
Move distance_initial_visibility to scenario setting

RenechCDDA committed:
3D FOV range is a game constant, not an option

RenechCDDA committed:
Remove more world options from the UI
-region_default ---> external
SEASON_LENGTH --> external
CONSTRUCTION_SCALING ---> external
ETERNAL_SEASON ----> external
-Removed the time group since it only had a single option left(ETERNAL_TIME_OF_DAY), moved that option to its own listing

Procyonae commented:
I don't think I agree with all of this but ignoring that for a minute do removed options need any kind of config migration or does it just silently ignore the non existent ones?

RenechCDDA commented:
> I don't think I agree with all of this but ignoring that for a minute do removed options need any kind of config migration or does it just silently ignore the non existent ones?

Moving them to externals actually causes the new value to load smoothly from what I can tell, although more testing is definitely needed. There are a lot of ways for an option to be loaded and it's a bit of a mess.

RenechCDDA committed:
Hide world options that are modified by sliders

RenechCDDA commented:
Due to implementation I have once again changed course. (Believe me this is not fun for me either)

None of the sliders have been removed. However, all of the slider-relevant options (including `MONSTER_UPGRADE_FACTOR` which I erroneously believed was not affected by the difficulty slider) have been permanently hidden from the options menu. They cannot be loaded as externals, as the world options take precedence. (I tried.) As far as I can tell, the sliders still set the options. The sliders are the *only* way to set their options.

One of the upsides to this is that it shouldn't* have any impact on existing savegames, which will continue to use the same world options at the same values as before.

There are downsides to this approach.

-We are still sending the option text to be translated, despite the fact it will never be displayed.

-Because they remain as world options, they cannot be overridden with externals.


*Note that the following options were not slider-based and may need migration for existing save games.
`region_default` - the external here should be the only one anyone was ever using, so probably fine?
`SEASON_LENGTH`/CONSTRUCTION_SCALING/ETERNAL_SEASON - existing save games will need to add an external if they were using non-default values.

I have given some thought about how to do this, and [broached the topic on the development discord. ](https://discord.com/channels/598523535169945603/598535827169083403/1332153519108788396) The consensus answer seems to be enabling a newly-created per-save unique mod on legacy saves with the externals defined at their old values. There are issues with this, like what if the option is removed/changed in the future?

RenechCDDA commented:
Worth noting the churn in 4876d8a975c460ee9bbe1413405aaf60ffa14fea is somewhat necessary. There's a fun bug with option groups: If all the options are `COPT_ALWAYS_HIDE` it still displays the first option, always.

RenechCDDA committed:
Nix the translations

RenechCDDA committed:
Actually move mutation by radiation to external

RenechCDDA committed:
Move season_lenght, construction_scaling, and eternal_season back to options.......

RenechCDDA commented:
This PR's gotten stale so I've polished off what work has done and postponed the rest. Specifically SEASON_LENGTH, CONSTRUCTION_SCALING, and ETERNAL_SEASON are all staying options for now, because I don't have the knowledge/willpower to write custom migrations for them.

RenechCDDA committed:
Merge branch 'master' into remove_world_options

RenechCDDA commented:
There is a very nasty-looking merge commit in here since #80438 was merged while this was stale.

I am not entirely certain that my commit plays nice with it, and I'm not sure how to test it either...

akrieger committed:
Remove MONSTER_UPGRADE_FACTOR world option.

Co-authored-by: RenechCDDA <84619419+RenechCDDA@users.noreply.github.com>

RenechCDDA committed:
Update includes for fov_3d_z_range

(force IWYU to keep cached_options.h in ranged.cpp for external int test_mode)

RenechCDDA commented:
There's some confusion over the final state of this PR, so to summarize:

Moved to scenarios:
Distance initial visibility

Moved to Externals:
Default region
Mutation by radiation

Removed as an option, completely:
3D FOV range

Untouched but planned for future move to externals:
Season length
Construction scaling
Eternal season

Hidden from the options menu, specific settings still available via the world sliders*:
City size
City spacing
Spawn density
Item spawn rate
NPC spawn rate
Monster speed scaling
Monster HP scaling


It is not possible to override these world options with externals, the option handling is "extremely convoluted" and I do not have a perfect grasp of how it works. We have several different types of options, all of which may be returned by get_option() and may all be used more or less interchangeably.

PatrikLundell commented:
I missed this, as I thought it was only about sliders, which I don't care about, but apparently it also removes essentially all advanced control altogether, leaving ONLY some nerfed sliders.

This goes against the "recent" trend in games, where settings have been complemented with advanced options to allow players to tailor their experience to their liking, rather than having to accept more or less intelligently assembled bundles.

To add insult to injury, when removing fine control of city sizes, the reduced resolution slider retains the quest breaking option of "rural" which at least breaks the Refugee Center blood analysis quest, as hospitals can't be generated if the city size is less than 6 (the eminent logic seems to be that world gen city size is half the JSON CITY_SIZES values), as hospitals require a min city size of 12. See #80528.

I accept that city sizes shouldn't be set to something less than what game quests require (and they fail to do), but see no reason for why players mustn't be allowed to exercise detailed control within the functioning slider range.

I suspect tight city spacing can make it impossible to generate at least the Gun Show quest for the artisans, and so that there might be a need to set a lower limit on that value, but, again, I don't see the logic behind hiding the exact values behind sliders (again, within the working range of values).

What's game breaking about adjusting zombie evolution rate, speed, strength, etc.?

I know there have been issues with reduced item spawns causing quest issues with quest items not spawning when they were not marked as quest items, but is that a sufficient ground to deny player control over their game play? And I don't even know if the slider does keep item clutter constant or not...

In general, hiding existing functionality behind composite sliders ("difficulty") in bundles somebody think are the perfect ones for everybody is making the game worse, not better, and forcing dissenters to make "mods" just to access the fine tuning that's now buried is not a player friendly approach.

I would suggest reverting this PR.

Maddremor commented:
I think the sliders aren't great either. The broader issue is that the game doesn't work well with the worldgen options we have provided up until now. The benefit of moving them to JSON/mods is that there would be some level of curation or DIY ethos to the changes. I'm not the one making the call, but I suspect that we will see the sliders also be curtailed in the future.

PatrikLundell commented:
The "sliders" aren't actually sliders, but just a small number of selections presented as if the selections were on a slider. There's nothing wrong with presenting the players with a limited number of standardized options, as it makes it easier to get started, but 5 or 10 years ago game developers started to realize that they could also provide the ability to customize the settings hidden behind these selections. I see no reason for C:DDA to go in the opposite direction.

Maddremor commented:
My point was rather that these presets should be mods, and if mapgen breaks (due to there not being enough space in a megacity configuration or no town being large enough in a rural one) the mod could manually tweak generation rules, as opposed to vanilla having to deal with any permutation or just breaking with no real recourse.

PatrikLundell commented:
Well, mods can really only be used if they're both supported and accepted by the main developers, and both are doubtful. Also, you'd get a rather large number of mods if each would tweak one parameter, and an even larger number if they'd contain the particular set their authors find suitable.
Also, who's going to do the support for things breaking due to combination issues?

Well, I guess there could be an "advanced settings" mod that restores the removed functionality...

Zireael07 commented:
Seconding @PatrikLundell that the number of mods/combos would be unmaintainable and breaking.

There surely is a way to disable certain missions based on world settings? If there isn't, there should be.

RenechCDDA commented:
> There surely is a way to disable certain missions based on world settings? If there isn't, there should be.

This requires us to be aware of all the possible setting combinations which can cause issues and then patch our json content around them. And also we'll somehow have to communicate that to the player and write exceptions to mission chains etc to handle all that. And then we'd have to filter all existing json around it, and somehow advise new contributors to consider every one of many possible option combinations lest their content be broken in some or all of them.

That is not a viable solution.

PatrikLundell commented:
As @RenechCDDA' said, working around missions would require a lot of work. A proper approach ought to be to decide on the range of settings the game supports and design missions around that. It would, of course, help a fair bit if the constraints and their consequences are communicated.

Until #80528 I had no idea the zombie blood mission was breakable by city settings, and I certainly hadn't any inkling on how the large numbers in CITY_SIZES related to the worldgen city settings. That's info that ought to be communicated to quest designers, and there's probably more.

The Gun Show is apparently problematic and probably should be adjusted by someone who understands why: Somehow the requirement to be at least 3 tiles outside of a city on a field 5 tiles from a road causes trouble. Would it be easier if the distance could be somewhat variable (if that's supported)? Is there an issue with (lack of) rotation that might restrict the road location to only be to the north?
